File: src/main/resources/templates/Autenticar.js
function generarLlaveSecreta() {
    var randomArray = new Uint8Array(20);
    window.crypto.getRandomValues(randomArray);
    var base32 = new base32Encode(randomArray);
    return base32;
}

function obtenerCodigoGoogle(llaveSecreta, cuenta) {
    var empresa = "Jimaye";
    try {
        return "otpauth://totp/"
            + encodeURIComponent(empresa + ":" + cuenta).replace(/%2B/g, '%20')
            + "?secret=" + encodeURIComponent(llaveSecreta).replace(/%2B/g, '%20')
            + "&issuer=" + encodeURIComponent(empresa).replace(/%2B/g, '%20');
    } catch (e) {
        throw new Error(e);
    }
}

function crearQR(codigoDatos) {
    var qrCode = new QRCodeStyling({
        width: 500,
        height: 500,
        data: codigoDatos,
        imageOptions: {
            hideBackgroundDots: true,
            imageSize: 0.8
        }
    });

    qrCode.append(document.getElementById("contenedorQR"));
}

function obtenerCodigoMostrado(llaveSecreta) {
    var base32 = new base32Decode(llaveSecreta);
    var hexKey = new hexEncode(base32);
    return TOTP.getOTP(hexKey);
}

File: src/main/java/com.proyectoDiseno/Authentication/Autenticar.java
package com.proyectoDiseno.Authentication;
import com.google.zxing.BarcodeFormat;
import com.google.zxing.MultiFormatWriter;
import com.google.zxing.WriterException;
import com.google.zxing.client.j2se.MatrixToImageWriter;
import com.google.zxing.common.BitMatrix;
import de.taimos.totp.TOTP;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.security.SecureRandom;
import org.apache.commons.codec.binary.Base32;
import org.apache.commons.codec.binary.Hex;

public class Autenticar {
    public static String generarLlaveSecreta() {
        SecureRandom random = new SecureRandom();
        byte[] bytes = new byte[20];
        random.nextBytes(bytes);
        Base32 base32 = new Base32();
        return base32.encodeToString(bytes);
    }

    public static String ObtenerCodigoGoogle(String llaveSecreta, String cuenta){
        String empresa = "Jimaye";
        try {
            return "otpauth://totp/"
                    + URLEncoder.encode(empresa + ":" + cuenta, "UTF-8").replace("+", "%20")
                    + "?secret=" + URLEncoder.encode(llaveSecreta, "UTF-8").replace("+", "%20")
                    + "&issuer=" + URLEncoder.encode(empresa, "UTF-8").replace("+", "%20");
        } catch (UnsupportedEncodingException e) {
            throw new IllegalStateException(e);
        }
    }

    public static void crearQR(String codigoDatos) throws WriterException, IOException {
        // Ruta relativa dentro del proyecto donde se guardará la imagen del código QR
        String rutaRelativa = "src/main/resources/templates/codigo.png";

        // Obtener la ruta absoluta del proyecto
        String rutaProyecto = System.getProperty("user.dir");
        String rutaCompleta = rutaProyecto + File.separator + rutaRelativa;

        BitMatrix matrix = new MultiFormatWriter().encode(codigoDatos, BarcodeFormat.QR_CODE,
                500, 500);
        try (FileOutputStream out = new FileOutputStream(rutaCompleta)) {
            MatrixToImageWriter.writeToStream(matrix, "png", out);
        }
    }

    public static String obtenerCodigoMostrado(String llaveSecreta) {
        Base32 base32 = new Base32();
        byte[] bytes = base32.decode(llaveSecreta);
        String hexKey = Hex.encodeHexString(bytes);
        return TOTP.getOTP(hexKey);
    }
}